### PR TITLE
Add warning message for items with the same name as Rust primitive types.

### LINKED ISSUE #1382: Better warnings for types with the same name as rust types
### Input C/C++ Header
See invocation below

### Bindgen Invocation

```Rust
    bindgen::builder()
        .header_contents("bug.h", "#include <stdint.h>\ntypedef uint64_t u64;")
        .generate()
        .unwrap()
```

### Actual Results

```
   Compiling bindgen-typedef-primitive v0.1.0 (file:///home/jon/dev/tmp/bindgen-typedef-primitive)
error[E0391]: cycle detected when processing `u64`
   --> /home/jon/dev/tmp/cargo-target/debug/build/bindgen-typedef-primitive-177200f3c53a8948/out/bindings.rs:193:16
    |
193 | pub type u64 = u64;
    |                ^^^
    |
    = note: ...which again requires processing `u64`, completing the cycle

error: aborting due to previous error
```

### Expected Results

This error message is rather confusing, especially when the header file for which bindings are being generated is large.
Rather, it would be nice if generating bindings for types which have the same name as rust types failed.

cc @jonhoo

### PR DESCRIPTION
**Pull request by kellda**
Fixes #1382

Mangling colliding items may break backwards compatibility
with handling of C/C++ structs/classes which have *fields* named u32,
f64, etc., as having fields with such names is valid Rust.

We would need `ctx.rust_mangle` to differentiate between field and type
names, and at the moment I am not sure how to approach that.

r? @emilio

### CODE REVIEW

File: src/ir/context.rs [hunk 1]
```diff
@@ -865,6 +865,27 @@ If you encounter an error missing from this list, please file an issue or a PR!"
...
+                "i128" |
+                "u64" |
+                "i64" |
+                "u32" |
+                "i32" |
+                "u16" |
+                "i16" |
+                "u8" |
+                "i8" => {
+                    if !self.options.allow_primitive_collisions {
```
Comment (emilio) [inline src/ir/context.rs]: I don't think this option is particularly useful if its only side-effect is `error!`-ing out. Having an option that preserves the current behavior may be worth it, but probably not.

So I think it's ok to do this minor breaking change. If somebody had a use-case for the pre-existing behavior we can add such an option.

Comment (kellda) [reply src/ir/context.rs]: Well yet you are referring to the previous implementation by @jhod0 that I rebased as a starting point. I then changed it to warn! by default and mangle when the flag is set.

**Comment by kellda** [comment]
The flag allows to avoid mangling to keep backward compatibility, but I don't think much crates are affected.

I'm not sure where / how to add a test, I'll see if I find.

**Comment by emilio** [comment]
Ah, that is true, I was reviewing the first commit. Anyhow I'm still on the fence about whether the flag is useful. I think I'd much rather keep the code simpler for now until somebody needs the flag (in which case we implement it and test it separately).

**Comment by kellda** [comment]
I removed flags, now the only change is adding Rust primitives to the list of things to be mangled.

About the test:
- Is there a C type that generates Rust's `f32`, `f64`, `i128`, `u128`, `bool` or `str`/`&str` ?
- How to find out why bindgen can't generate bindings ?

**Comment by kellda** [comment]
I found the keywords test. Is it enough to test this ? Or should I  add tests with primitives used in typedefs and structs ? If so, in the keywords test or in another one ?

**Comment by kellda** [comment]
Should be good now

Btw I got some idea while writing tests: is generating e.g. `core::primitive::u64` instead of just `u64` a possible / good way to solve this instead of mangling items ?

**Comment by kellda** [comment]
>I wasn't aware of the types being exported under core...

Err, in fact they are only exported under std. My bad

**Comment by emilio** [comment]
Ok, let's do this for now then. Thank you!

**Comment by emilio** [comment]
Yes, thanks. I shouldn't have missed this. I guess when skimming the commit log to make a new release I didn't realize that there were cases this patch changed that compiled before.

**Comment by emilio** [comment]
0.54.1 is yanked, will publish 0.55 asap (need to go through all changes and write the changelog). Sorry for the breakage @khuey

**Pull request merged**